README.md
[![Build status](https://img.shields.io/github/workflow/status/belm0/perf-timer/CI)](https://github.com/belm0/perf-timer/actions/workflows/ci.yml?query=branch%3Amaster+)
[![Package version](https://img.shields.io/pypi/v/perf-timer.svg)](https://pypi.org/project/perf-timer)
[![Supported Python versions](https://img.shields.io/pypi/pyversions/perf-timer.svg)](https://pypi.org/project/perf-timer)

# PerfTimer

An indispensable performance timer for Python

## Background
### Taxonomy
Three general tools should be employed to
understand the CPU performance of your Python code:
  1. **sampling profiler** - measures the relative
  distribution of time spent among function or
  lines of code during a program session.  Limited by
  sampling resolution.  Does not provide call counts,
  and results cannot be easily compared between sessions.
  2. **microbenchmark timer** (timeit) - accurately
  times a contrived code snippet by running it repeatedly
  3. **instrumenting timer** - accurately times a specific
  function or section of your code during a program
  session

_PerfTimer_ is a humble instance of #3.  It's the easiest
way (least amount of fuss and effort) to get insight into
call count and execution time of a function or piece
of code during a real session of your program.

Use cases include:
  * check the effects of algorithm tweaks, new implementations, etc.
  * confirm the performance of a library you are considering under
  actual use by your app (as opposed to upstream's artificial
  benchmarks)
  * measure CPU overhead of networking or other asynchronous I/O
  (currently supported: OS threads, Trio async/await)

### Yet another code timer?

It seems everyone has tried their hand at writing one of these timer
utilities.  Implementations can be found in public repos, snippets, and PyPi—
there's even a Python feature request.  That's not counting all the
proprietary and one-off instances.

Features of this library:

  * **flexible** - use as a context manager or function decorator;
  pluggable logging, timer, and observer functions
  * **low overhead** (typically a few microseconds) - can be
  employed in hot code paths or even enabled on production deployments
  * **async/await support** (Trio only) - first of its kind!  Periods when a task is
  is sleeping, blocked by I/O, etc. will not be counted.
  * **percentile durations** - e.g. report the median and 90th percentile
  execution time of the instrumented code.  Implemented with a bounded-memory,
  streaming histogram.

## Usage

Typical usage is to create a `PerfTimer` instance at the global
scope, so that aggregate execution time is reported at program termination:

```python
from perf_timer import PerfTimer

_timer = PerfTimer('process thumbnail')

def get_thumbnail_image(path):
    img = cache.get_thumbnail(path)
    if not thumbnail:
        img = read_image(path)
        with _timer:
            img.decode()
            img.resize(THUMBNAIL_SIZE)
        cache.set_thumbnail(img)
    return img
```

When the program exits, assuming `get_thumbnail_image` was called
several times, execution stats will be reported to stdout as
follows:

```
timer "process thumbnail": avg 73.1 µs ± 18.0 µs, max 320.5 µs in 292 runs
```

### decorator style

To instrument an entire function or class method, use `PerfTimer`
as a decorator:

```python
@PerfTimer('get thumbnail')
def get_thumbnail_image(path):
    ...
```

### histogram statistics

By default `PerfTimer` will track the average, standard deviation, and maximum
of observed values.  Other available observers include `HistogramObserver`,
which reports (customizable) percentiles:

```python
import random
import time
from perf_timer import PerfTimer, HistogramObserver

_timer = PerfTimer('test', observer=HistogramObserver, quantiles=(.5, .9))
for _ in range(50):
    with _timer:
        time.sleep(random.expovariate(1/.1))

del _timer
```
output:
```
timer "test": avg 117ms ± 128ms, 50% ≤ 81.9ms, 90% ≤ 243ms in 50 runs
```

### custom logging

A custom logging function may be passed to the `PerfTimer`
constructor:

```python
import logging

_logger = logging.getLogger()
_timer = PerfTimer('process thumbnail', log_fn=_logger.debug)
```

### OS thread support

To minimize overhead, `PerfTimer` assumes single-thread access.  Use
`ThreadPerfTimer` in multi-thread scenarios:

```python
from perf_timer import ThreadPerfTimer

_timer = ThreadPerfTimer('process thumbnail')
```

### async support

In the previous example, timing the entire function will include file
I/O time since `PerfTimer` measures wall time by default.  For programs
which happen to do I/O via the Trio async/await library, you
can use `TrioPerfTimer` which measures time only when the current task
is executing:

```python
from perf_timer import TrioPerfTimer

@TrioPerfTimer('get thumbnail')
async def get_thumbnail_image(path):
    img = cache.get_thumbnail(path)
    if not thumbnail:
        img = await read_image(path)
        img.decode()
        img.resize(THUMBNAIL_SIZE)
        cache.set_thumbnail(img)
    return img
```

(Open challenge: support other async/await libraries)

### trio_perf_counter()

This module also provides the `trio_perf_counter()` primitive.
Following the semantics of the various performance counters in Python's `time`
module, `trio_perf_counter()` provides high resolution measurement of a Trio
task's execution time, excluding periods where it's sleeping or blocked on I/O.
(`TrioPerfTimer` uses this internally.)

```python
from perf_timer import trio_perf_counter

async def get_remote_object():
    t0 = trio_perf_counter()
    msg = await read_network_bytes()
    obj = parse(msg)
    print('task CPU usage (seconds):', trio_perf_counter() - t0)
    return obj
```

## Installation

```shell
pip install perf-timer
```

## Measurement overhead

Measurement overhead is important.  The smaller the timer's overhead, the
less it interferes with the normal timing of your program, and the tighter
the code loop it can be applied to.

The values below represent the typical overhead of one observation, as measured
on ye old laptop (2014 MacBook Air 11 1.7GHz i7).

```
$ pip install -r test-requirements.txt
$ python benchmarks/overhead.py
compare observers:
    PerfTimer(observer=AverageObserver):         1.5 µs
    PerfTimer(observer=StdDevObserver):          1.8 µs  (default)
    PerfTimer(observer=HistogramObserver):       6.0 µs

compare types:
    PerfTimer(observer=StdDevObserver):          1.8 µs
    ThreadPerfTimer(observer=StdDevObserver):    9.8 µs
    TrioPerfTimer(observer=StdDevObserver):      4.8 µs
```

## TODO
  * features
    * faster HistogramObserver
    * more async/await support: asyncio, curio, etc.
      * [asyncio hint which no longer works](https://stackoverflow.com/revisions/34827291/3)
  * project infrastructure
    * code coverage integration
    * publish docs
    * type annotations and check


CHANGELOG.md
# Release history

## perf-timer 0.3.0 (pending)
### Fixed
- fix `__del__` exception on badly-constructed instances

## perf-timer 0.2.2 (2021-03-02)
### Fixed
- handle absence of `time.thread_timer()` gracefully.  This timer, which is the
  default used by `ThreadPerfTimer`, may not be available in some OS X
  environments.

## perf-timer 0.2.1 (2020-11-09)
### Fixed
- employ `atexit()` to robustly log results even when `__del__` finalizers are
  not called

## perf-timer 0.2.0 (2020-07-01)
### Added
- perf-timer classes now support tracking various statistics
  including standard deviation and percentiles.  The options are
  `AverageObserver`, `StdDevObserver` (default), and `HistogramObserver`.
  E.g. `PerfTimer(..., observer=HistogramObserver)`.
- Benchmark overhead of the various observer and timer types

## perf-timer 0.1.1 (2020-06-05)
### Fixed
- Support rename of trio.hazmat to trio.lowlevel
- Expose docs to help()

## perf-timer 0.1.0 (2019-07-31)
Initial version


tests/test_trio.py
import time
from unittest.mock import Mock

import pytest
import trio
try:
    import trio.lowlevel as trio_lowlevel
except ImportError:
    import trio.hazmat as trio_lowlevel

from perf_timer import trio_perf_counter, _trio, TrioPerfTimer, AverageObserver, trio_hierarchy_perf_counter


async def _work(duration, count=1):
    for _ in range(count):
        time.sleep(duration)
        await trio.sleep(0)


async def test_descheduled_time_instrument():
    time_fn = Mock(side_effect=[5, 10, 10, 20])
    instrument = _trio._DescheduledTimeInstrument(time_fn=time_fn)
    trio_lowlevel.add_instrument(instrument)

    # Only tasks referenced by get_elapsed_descheduled_time() will be tracked,
    # so instrument is not tracking the current task.
    await trio.sleep(0)
    assert not time_fn.called

    async with trio.open_nursery() as nursery:
        @nursery.start_soon
        async def _tracked_child():
            # calling get_elapsed_descheduled_time() initiates tracking
            task = trio_lowlevel.current_task()
            assert instrument.get_elapsed_descheduled_time(task) == 0
            await trio.sleep(0)
            assert instrument.get_elapsed_descheduled_time(task) == 10 - 5
            await trio.sleep(0)
            assert instrument.get_elapsed_descheduled_time(task) == 20 - 5
            # time function is called twice for each deschedule
            assert time_fn.call_count == 4

    # the sole tracked task exited, so instrument is automatically removed
    with pytest.raises(KeyError):
        trio_lowlevel.remove_instrument(instrument)


async def test_descheduled_time_instrument_exclude_children():
    time_fn = Mock(side_effect=[5, 10])
    instrument = _trio._DescheduledTimeInstrument(time_fn=time_fn)
    trio_lowlevel.add_instrument(instrument)

    task = trio_lowlevel.current_task()
    assert instrument.get_elapsed_descheduled_time(task) == 0

    async with trio.open_nursery() as nursery:
        @nursery.start_soon
        async def _untracked_child():
            await trio.sleep(0)

    assert instrument.get_elapsed_descheduled_time(task) == 10 - 5
    assert time_fn.call_count == 2  # 2 x 1 deschedule (due to nursery)

    # our task is still alive, so instrument remains active
    trio_lowlevel.remove_instrument(instrument)


async def test_trio_perf_counter_time_sleep():
    # NOTE: subject to false pass due to reliance on wall time
    t0 = trio_perf_counter()
    await _work(.1, 5)
    dt = trio_perf_counter() - t0
    assert dt == pytest.approx(.5, rel=.15)


async def test_trio_perf_counter_child():
    t0 = trio_hierarchy_perf_counter()
    async with trio.open_nursery() as nursery:
        await _work(.05, 3)

        @nursery.start_soon
        async def _child():
            async with trio.open_nursery() as nursery2:
                @nursery2.start_soon
                async def _child_child():
                    await _work(.05, 5)

                await _work(.05, 5)

        @nursery.start_soon
        async def _child_2():
            await _work(.05, 10)

    dt = trio_hierarchy_perf_counter() - t0
    assert dt == pytest.approx(1.15, rel=.15)


async def test_trio_hierarchy_perf_counter_active_children():
    # trio_hierarchy_perf_counter() should work even if the task already has children
    async with trio.open_nursery() as nursery:
        @nursery.start_soon
        async def _child():
            async with trio.open_nursery() as nursery2:
                @nursery2.start_soon
                async def _child_child():
                    await _work(.1, 5)

                await _work(.1, 5)

        await trio.sleep(.5)
        t0 = trio_hierarchy_perf_counter()

        @nursery.start_soon
        async def _child_2():
            await _work(.1, 10)

    dt = trio_hierarchy_perf_counter() - t0
    assert dt == pytest.approx(1.5, rel=.15)


@pytest.mark.parametrize('counter_fn, instrument', [
    (trio_perf_counter, _trio._instrument),
    (trio_hierarchy_perf_counter, _trio._hierarchy_instrument),
])
async def test_trio_perf_counter_unregister(counter_fn, instrument: trio.abc.Instrument):
    async def perf_counter_with_trio_sleep():
        counter_fn()
        await trio.sleep(0)
        counter_fn()

    async with trio.open_nursery() as nursery:
        nursery.start_soon(perf_counter_with_trio_sleep)
        nursery.start_soon(perf_counter_with_trio_sleep)

    # Since all tasks using task_perf_counter() have exited, we expected
    # the Trio instrumentation to no longer be active (so remove call
    # will fail).
    with pytest.raises(KeyError):
        trio_lowlevel.remove_instrument(instrument)


async def test_trio_perf_timer(autojump_clock):
    # time_fn is called on enter and exit of each with block
    time_fn = Mock(side_effect=[10, 15,
                                15, 25])
    timer = TrioPerfTimer('foo', observer=AverageObserver, time_fn=time_fn)

    for _ in range(2):
        with timer:
            await trio.sleep(1)

    assert timer._count == 2
    assert timer._sum == 15
    assert timer._max == 10
    del timer


async def test_trio_perf_timer_decorator(autojump_clock):
    time_fn = Mock(side_effect=[10, 15,
                                15, 25])
    timer = TrioPerfTimer('foo', observer=AverageObserver, time_fn=time_fn)

    @timer
    async def foo():
        await trio.sleep(1)

    for _ in range(2):
        await foo()

    assert timer._count == 2
    assert timer._sum == 15
    assert timer._max == 10
    del timer


# TODO: test nested perf timers on the same task
# async def _foo():
#     with TrioPerfTimer('timer1'):
#         async with trio.open_nursery() as nursery:
#             @nursery.start_soon
#             async def _child_1():
#                 pass  # do work
#
#             # do work
#
#             with TrioPerfTimer('timer2'):
#                 @nursery.start_soon
#                 async def _child_2():
#                     pass  # do work


tests/test_observer.py
import random
from unittest.mock import Mock

import numpy
import pytest

from perf_timer import AverageObserver, StdDevObserver, HistogramObserver


@pytest.mark.parametrize('n', (0, 1, 100))
def test_average_observer(n):
    name = 'foo'
    log_fn = Mock()
    observer = AverageObserver(name, log_fn=log_fn)
    points = []
    for _ in range(n):
        x = random.expovariate(1/5)  # expected average is 5s
        observer._observe(x)
        points.append(x)
    observer._report()
    if n > 1:
        # timer "foo": avg 11.9 ms, max 12.8 ms in 10 runs
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'avg {sum(points)/n:.2f} s, '
            f'max {max(points):.1f} s '
            f'in {n} runs')
    elif n > 0:
        # timer "foo": 12.8 ms
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'{points[0]:.1f} s')
    else:
        log_fn.assert_not_called()


@pytest.mark.parametrize('n', (0, 1, 100))
def test_std_dev_observer(n):
    name = 'foo'
    log_fn = Mock()
    observer = StdDevObserver(name, log_fn=log_fn)
    points = []
    for _ in range(n):
        x = random.expovariate(1/5)  # expected average is 5s
        observer._observe(x)
        points.append(x)
    observer._report()
    if n > 1:
        # timer "foo": avg 11.9 ms ± 961 µs, max 12.8 ms in 10 runs
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'avg {sum(points)/n:.2f} s ± {numpy.std(points):.2f} s, '
            f'max {max(points):.1f} s '
            f'in {n} runs')
    elif n > 0:
        # timer "foo": 12.8 ms
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'{points[0]:.1f} s')
    else:
        log_fn.assert_not_called()


@pytest.mark.parametrize('n', (0, 1, 50))
def test_histogram_observer(n):
    name = 'foo'
    # cheating to allow a simple string compare:
    #   * requested quantiles and distribution are such that output precision is fixed
    #   * since n < max bins of the approximate histogram, quantile output will be exact
    quantiles = (.4, .5, .6)
    log_fn = Mock()
    observer = HistogramObserver(name, quantiles=quantiles, log_fn=log_fn)
    points = []
    for _ in range(n):
        x = random.expovariate(1/5)  # expected average is 5s
        observer._observe(x)
        points.append(x)
    observer._report()
    if n > 1:
        q_expected = numpy.quantile(points, quantiles)
        # timer "foo": avg 11.9ms ± 961µs, 50% ≤ 12.6ms, 90% ≤ 12.7ms in 10 runs
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'avg {sum(points)/n:.2f}s ± {numpy.std(points):.2f}s, '
            f'{", ".join(f"{q:.0%} ≤ {out:.2f}s" for q, out in zip(quantiles, q_expected))} '
            f'in {n} runs')
    elif n > 0:
        # timer "foo": 12.8 ms
        log_fn.assert_called_once_with(
            f'timer "{name}": '
            f'{points[0] * 1000:.1f} ms')
    else:
        log_fn.assert_not_called()


def test_histogram_observer_bad_input():
    with pytest.raises(ValueError):
        HistogramObserver('foo', quantiles=(.5, 2))
    with pytest.raises(ValueError):
        HistogramObserver('foo', quantiles=(.6, .5))


tests/test_histogram.py
import random

import numpy as np
import pytest
from pytest import approx

from perf_timer._histogram import ApproximateHistogram


def test_histogram_exact():
    random.seed(0)
    max_bins = 50
    h = ApproximateHistogram(max_bins=max_bins)
    points = []

    for _ in range(max_bins):
        p = random.expovariate(1/5)
        points.append(p)
        h.add(p)

    q = [i / 100 for i in range(101)]
    assert h.quantile(q) == approx(np.quantile(points, q))
    assert h.mean() == approx(np.mean(points))
    assert h.std() == approx(np.std(points))
    assert h.sum() == approx(np.sum(points))
    assert h.min == min(points)
    assert h.max == max(points)
    assert h.count == max_bins


@pytest.mark.parametrize("max_bins,num_points,expected_error", [
    (50, 50, 1e-6),
    (100, 150, 1.5),
    (100, 1000, 1),
    (250, 1000, .5),
])
def test_histogram_approx(max_bins, num_points, expected_error):
    random.seed(0)
    h = ApproximateHistogram(max_bins=max_bins)
    points = []

    for _ in range(num_points):
        p = random.expovariate(1/5)
        points.append(p)
        h.add(p)

    q = [i / 100 for i in range(101)]
    err_sum = 0  # avg percent error across samples
    for p, b, b_np, b_np_min, b_np_max in zip(
            q,
            h.quantile(q),
            np.quantile(points, q),
            np.quantile(points, [0] * 7 + q),
            np.quantile(points, q[7:] + [1] * 7)):
        err_denom = b_np_max - b_np_min
        err_sum += abs(b - b_np) / err_denom
    assert err_sum <= expected_error
    assert h.mean() == approx(np.mean(points))
    assert h.std() == approx(np.std(points), rel=.05)
    assert h.sum() == approx(np.sum(points))
    assert h.min == min(points)
    assert h.max == max(points)
    assert h.count == num_points


benchmarks/overhead.py
"""Measure and report overhead of the perf-timer variants.

The typical observation duration is reported for each case.

Synopsis:
    $ python overhead.py
    compare observers:
        PerfTimer(observer=AverageObserver):         1.5 µs
        PerfTimer(observer=StdDevObserver):          1.8 µs  (default)
        PerfTimer(observer=HistogramObserver):       6.0 µs

    compare types:
        PerfTimer(observer=StdDevObserver):          1.8 µs
        ThreadPerfTimer(observer=StdDevObserver):    9.8 µs
        TrioPerfTimer(observer=StdDevObserver):      4.8 µs
"""

from functools import partial

import trio

from perf_timer import (PerfTimer, ThreadPerfTimer, TrioPerfTimer,
                        AverageObserver, StdDevObserver, HistogramObserver,
                        measure_overhead, TrioHierarchyPerfTimer)
from perf_timer._impl import _format_duration


async def main():
    _format = partial(_format_duration, precision=2)
    default_observer = StdDevObserver
    print('compare observers:')
    timer_type = PerfTimer
    for observer in (AverageObserver, StdDevObserver, HistogramObserver):
        duration = measure_overhead(partial(timer_type, observer=observer))
        item = f'{timer_type.__name__}(observer={observer.__name__}):'
        print(f'    {item:45s}{_format(duration)}'
              f'{"  (default)" if observer is default_observer else ""}')

    print()
    print('compare types:')
    observer = default_observer
    for timer_type in (PerfTimer, ThreadPerfTimer, TrioPerfTimer, TrioHierarchyPerfTimer):
        duration = measure_overhead(partial(timer_type, observer=observer))
        item = f'{timer_type.__name__}(observer={observer.__name__}):'
        print(f'    {item:45s}{_format(duration)}')


if __name__ == '__main__':
    trio.run(main)


src/perf_timer/_impl.py
import atexit
import functools
import math
import timeit
from weakref import WeakSet

from contextvars import ContextVar
from inspect import iscoroutinefunction
from multiprocessing import Lock
from time import perf_counter
try:
    from time import thread_time
except ImportError:
    # thread_time is not available in some OS X environments
    thread_time = None

from perf_timer._histogram import ApproximateHistogram

_start_time_by_instance = ContextVar('start_time', default={})
_timers = WeakSet()


def _format_duration(duration, precision=3, delimiter=' '):
    """Returns human readable duration.

    >>> _format_duration(.0507)
    '50.7 ms'
    """
    units = (('s', 1), ('ms', 1e3), ('µs', 1e6), ('ns', 1e9))
    i = len(units) - 1
    if duration > 0:
        i = min(-int(math.floor(math.log10(duration)) // 3), i)
    symbol, scale = units[i]
    # constant precision, keeping trailing zeros but don't end in decimal point
    value = f'{duration * scale:#.{precision}g}'.rstrip('.')
    return f'{value}{delimiter}{symbol}'


class _BetterContextDecorator:
    """
    Equivalent to contextlib.ContextDecorator but supports decorating async
    functions.  The context manager itself it still non-async.
    """

    def _recreate_cm(self):
        return self

    def __call__(self, func):
        if iscoroutinefunction(func):
            @functools.wraps(func)
            async def inner(*args, **kwargs):
                with self._recreate_cm():  # pylint: disable=not-context-manager
                    return await func(*args, **kwargs)
        else:
            @functools.wraps(func)
            def inner(*args, **kwargs):
                with self._recreate_cm():  # pylint: disable=not-context-manager
                    return func(*args, **kwargs)
        return inner


class _PerfTimerBase(_BetterContextDecorator):

    # NOTE: `observer` is handled by the metaclass, and `quantiles` is handled
    #   by HistogramObserver.  They're included here only for documentation.
    def __init__(self, name, *, time_fn=perf_counter, log_fn=print,
                 observer=None, quantiles=None):
        """
        :param name: string used to annotate the timer output
        :param time_fn: optional function which returns the current time.
            (A None value will raise NotImplementedError.)
        :param log_fn: optional function which records the output string
        :param observer: mixin class to observe and summarize samples
            (AverageObserver|StdDevObserver|HistogramObserver, default StdDevObserver)
        :param quantiles: for HistogramObserver, a sequence of quantiles to report.
            Values must be in range [0..1] and monotonically increasing.
            (default: (0.5, 0.9, 0.98))
        """
        self._init_ok = False
        if not time_fn:
            raise NotImplementedError
        self.name = name
        self._time_fn = time_fn
        self._log_fn = log_fn
        self._startTimeByInstance = _start_time_by_instance.get()
        self._reported = False
        _timers.add(self)
        self._init_ok = True

    def _observe(self, duration):
        """called for each observed duration"""

    def _report(self):
        """called to report observation results"""

    def _report_once(self):
        if not self._reported:
            self._report()
            self._reported = True

    def __del__(self):
        if self._init_ok:
            self._report_once()

    def __enter__(self):
        if self in self._startTimeByInstance:
            raise RuntimeError('PerfTimer is not re-entrant')
        self._startTimeByInstance[self] = self._time_fn()

    def __exit__(self, exc_type, exc_value, traceback):
        current_time = self._time_fn()
        start_time = self._startTimeByInstance.pop(self)
        if exc_type is None:
            duration = current_time - start_time
            self._observe(duration)


class AverageObserver(_PerfTimerBase):
    """Mixin which outputs mean and max

    output synopsis:
        timer "foo": avg 11.9 ms, max 12.8 ms in 10 runs
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._count = 0
        self._sum = 0
        self._max = -math.inf

    def _observe(self, duration):
        self._count += 1
        self._sum += duration
        self._max = max(self._max, duration)

    def _report(self):
        if self._count > 1:
            mean = self._sum / self._count
            self._log_fn(f'timer "{self.name}": '
                         f'avg {_format_duration(mean)}, '
                         f'max {_format_duration(self._max)} '
                         f'in {self._count} runs')
        elif self._count > 0:
            self._log_fn(f'timer "{self.name}": '
                         f'{_format_duration(self._sum)}')


class StdDevObserver(_PerfTimerBase):
    """Mixin which outputs mean, stddev, and max

    15 - 20% slower than _AverageObserver.
    https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

    output synopsis:
        timer "foo": avg 11.9 ms ± 961 µs, max 12.8 ms in 10 runs
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._count = 0
        self._mean = 0
        self._m2 = 0
        self._max = -math.inf

    def _observe(self, duration):
        self._count += 1
        delta = duration - self._mean
        self._mean += delta / self._count
        self._m2 += delta * (duration - self._mean)
        self._max = max(self._max, duration)

    def _report(self):
        if self._count > 1:
            std = math.sqrt(self._m2 / self._count)
            self._log_fn(f'timer "{self.name}": '
                         f'avg {_format_duration(self._mean)} '
                         f'± {_format_duration(std)}, '
                         f'max {_format_duration(self._max)} '
                         f'in {self._count} runs')
        elif self._count > 0:
            self._log_fn(f'timer "{self.name}": '
                         f'{_format_duration(self._mean)}')


class HistogramObserver(_PerfTimerBase):
    """Mixin which outputs mean, standard deviation, and percentiles

    output synopsis:
        timer "foo": avg 11.9ms ± 961µs, 50% ≤ 12.6ms, 90% ≤ 12.7ms in 10 runs
    """

    def __init__(self, *args, quantiles=(.5, .9, .98), max_bins=64, **kwargs):
        super().__init__(*args, **kwargs)
        self._init_ok = False
        if not all(0 <= x <= 1 for x in quantiles):
            raise ValueError('quantile values must be in the range [0, 1]')
        if not all(a < b for a, b in zip(quantiles, quantiles[1:])):
            raise ValueError('quantiles must be monotonically increasing')
        self._quantiles = quantiles
        self._hist = ApproximateHistogram(max_bins=max_bins)
        self._init_ok = True

    def _observe(self, duration):
        self._hist.add(duration)

    def _report(self):
        if self._hist.count > 1:
            _format = functools.partial(_format_duration, delimiter='')
            hist_quantiles = self._hist.quantile(self._quantiles)
            percentiles = [f"{pct * 100:.0f}% ≤ {_format(val)}"
                           for pct, val in zip(self._quantiles, hist_quantiles)]
            self._log_fn(f'timer "{self.name}": '
                         f'avg {_format(self._hist.mean())} '
                         f'± {_format(self._hist.std())}, '
                         f'{", ".join(percentiles)} '
                         f'in {self._hist.count} runs')
        elif self._hist.count > 0:
            self._log_fn(f'timer "{self.name}": '
                         f'{_format_duration(self._hist.sum())}')


class _ObservationLock(_PerfTimerBase):
    """Mixin which wraps _observe() in a lock"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._lock = Lock()

    def _observe(self, duration):
        with self._lock:
            super()._observe(duration)


class _MixinMeta(type):
    """Metaclass which injects an observer mixin based on constructor arg"""

    @staticmethod
    @functools.lru_cache(maxsize=None)
    def _get_cls(observer, cls):
        # NOTE: bases ordering allows _ObservationLock to override the observer
        return type(cls.__name__, (cls, observer), {})

    def __call__(cls, *args, observer=StdDevObserver, **kwargs):
        out_cls = _MixinMeta._get_cls(observer, cls)
        return type.__call__(out_cls, *args, **kwargs)


class PerfTimer(_PerfTimerBase, metaclass=_MixinMeta):
    """Performance timer

    Use to measure performance of a block of code.  The object will log
    performance stats when it is destroyed.

        perf_timer = PerfTimer('my code')
        ...
        def foo():
            ...
            with perf_timer:
                # code under test
            ...

    It can also be used as a function decorator:

        @PerfTimer('my function')
        def foo():
            ...

    This implementation is not thread safe.  For a multi-threaded scenario,
    use `ThreadPerfTimer`.
    """


class ThreadPerfTimer(_ObservationLock, PerfTimer):
    """Variant of PerfTimer which measures CPU time of the current thread

    (Implemented with time.thread_time by default, which may not be available
    in some OS X environments.)
    """

    def __init__(self, name, time_fn=thread_time, **kwargs):
        super().__init__(name, time_fn=time_fn, **kwargs)


def measure_overhead(timer_factory):
    """Measure the overhead of a timer instance from the given factory.

    :param timer_factory: callable which returns a new timer instance
    :return: the average duration of one observation, in seconds
    """
    timeit_timer = timeit.Timer(
        globals={'timer': timer_factory('foo', log_fn=lambda x: x)},
        stmt='with timer: pass'
    )
    n, duration = timeit_timer.autorange()
    min_duration = min([duration] + timeit_timer.repeat(number=n))
    return min_duration / n


@atexit.register
def _atexit():
    while _timers:
        _timers.pop()._report_once()
